src/main/java/com/commercetools/sync/services/impl/CategoryServiceImpl.java

```diff
@@ -143,22 +145,30 @@ public CategoryServiceImpl(@Nonnull final BaseSyncOptions syncOptions) {
     @Nonnull
     @Override
     public CompletionStage<Optional<Category>> createCategory(@Nonnull final CategoryDraft categoryDraft) {
-        final CategoryCreateCommand categoryCreateCommand = CategoryCreateCommand.of(categoryDraft);
-        return syncOptions.getCtpClient().execute(categoryCreateCommand)
-                          .handle((createdCategory, sphereException) -> {
-                              // TODO: Refactor to reuse below duplicate code and ProductServiceImpl.
-                              if (sphereException != null) {
-                                  syncOptions
-                                      .applyErrorCallback(format(CREATE_FAILED, categoryDraft.getKey(),
-                                          sphereException), sphereException);
-                                  return Optional.empty();
-                              } else {
-                                  keyToIdCache.put(createdCategory.getKey(), createdCategory.getId());
-                                  return Optional.of(createdCategory);
-                              }
-                          });
+        return applyCallbackAndCreate(categoryDraft, syncOptions, CategoryCreateCommand::of,
+                this::handleCategoryCreation);
     }
 
+    private Optional<Category> handleCategoryCreation(
+            @Nonnull final CategoryDraft draft,
+            @Nullable final Category createdCategory,
+            @Nullable final Throwable sphereException) {
+
+        final Supplier<Optional<Category>> onFailure = () -> {
+            syncOptions.applyErrorCallback(format(CREATE_FAILED, draft.getKey(), sphereException), sphereException);
+            return Optional.empty();
+        };
+
+        @SuppressWarnings("ConstantConditions") //createdCategory is only invoked if not null.
+        final Supplier<Optional<Category>> onSuccess = () -> {
+            keyToIdCache.put(createdCategory.getKey(), createdCategory.getId());
+            return Optional.of(createdCategory);
+        };
+
+        return createdCategory != null ? onSuccess.get() : onFailure.get();
```

Comment:
looks like an overhead? You created two callback, called one of them immediately, and then forgot them both...
why not just pure old `if-then-else`

Reply:
no idea why I left with suppliers when with normal `if-then-else` would have been simpler. I was refactoring a lot around it yesterday, I didn't notice: 714ce79

Thanks for noticing it 😄